runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -2937,19 +2937,339 @@ J9::Z::TreeEvaluator::anewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg
+   generateRIEInstruction(cg, TR::InstOpCode::CLGIJ, node, firstDimLenReg, (int8_t)0, nonZeroFirstDimLabel, (TR::InstOpCode::COND_BNE));
+
+   // TODO: We don't need this label. But it's good for organizing the different paths for now.
+   TR::LabelSymbol *firstDimZeroLabel = generateLabelSymbol(cg);
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, firstDimZeroLabel);
+
+   // First Dim zero, only allocate 1 zero-length object array
+   int32_t zeroArraySize = TR::Compiler->om.discontiguousArrayHeaderSizeInBytes();
+   TR::Register *vmThreadReg = cg->getMethodMetaDataRealRegister();
+   generateRXInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, targetReg, generateS390MemoryReference(vmThreadReg, offsetof(J9VMThread, heapAlloc), cg));
+
+   // Branch to OOL if there's not enough space for an array of size 0.
+   TR::Register *temp1Reg = cg->allocateRegister();
+   generateRIEInstruction(cg, TR::InstOpCode::getAddHalfWordImmDistinctOperandOpCode(), node, temp1Reg, targetReg, zeroArraySize);
+   generateRXInstruction(cg, TR::InstOpCode::getCmpLogicalOpCode(), node, temp1Reg, generateS390MemoryReference(vmThreadReg, offsetof(J9VMThread, heapTop), cg));
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BH, node, oolFailLabel);
+
+   // If there's enough space, then we can continue to allocate.
+   generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), node, temp1Reg, generateS390MemoryReference(vmThreadReg, offsetof(J9VMThread, heapAlloc), cg));
+   
+   // Init class
+   bool use64BitClasses = comp->target().is64Bit() && !TR::Compiler->om.generateCompressedObjectHeaders();
+   generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), node, classReg, generateS390MemoryReference(targetReg, TR::Compiler->om.offsetOfObjectVftField(), cg));
+   
+   // Init size and '0' fields to 0
+   TR::Register *temp0Reg = cg->allocateRegister();
+   generateRRInstruction(cg, TR::InstOpCode::getXORRegOpCode(), node, temp0Reg, temp0Reg);
+   generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), node, temp0Reg, generateS390MemoryReference(targetReg, fej9->getOffsetOfContiguousArraySizeField(), cg));
+   generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), node, temp0Reg, generateS390MemoryReference(targetReg, fej9->getOffsetOfDiscontiguousArraySizeField(), cg));
+
+   // Done with this scenario, so jump to end of ICF.
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, cFlowRegionEnd);
+
+   // We end up in this block of the ICF if the first dimension is non-zero and the second dimension is zero.
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, nonZeroFirstDimLabel); // DCDCDCDC - > This is temporary. In the long run we want to have actual code and not just fall through into oolFailLabel.
```

Comment:
"DCDC" needs to be removed. If there is something to be done, we should leave a TODO and open an issue for it.